docs/architecture/dapr-for-net-developers/publish-subscribe-buildingblock.md

```diff
@@ -0,0 +1,398 @@
+
+```csharp
+public interface IEventBus
+{
+    void Publish(IntegrationEvent @event);
+
+    void Subscribe<T, TH>()
+        where T : IntegrationEvent
+        where TH : IIntegrationEventHandler<T>;
+}
+```
+
+Concrete implementations of this interface exist in eShopOnContainers for both RabbitMQ and Azure Service Bus. Each implementation included a great deal of custom plumbing code that was complex to understand and difficult to maintain.
+
+The newer eShopOnDapr significantly simplifies Publish/Subscribe behavior by using Dapr. For example, the `IEventBus` interface was reduced to a single method:
+
+```csharp
+public interface IEventBus
+{
+    Task PublishAsync(IntegrationEvent @event);
+}
+```
+
+### Publishing events
+
+In the updated eShopOnDapr, a single `DaprEventBus` implementation can support any Dapr-supported message broker. The following code block shows the simplified Publish method. Note how the `PublishAsync` method uses the Dapr client to publish an event:
+
+```csharp
+public class DaprEventBus : IEventBus
+{
+    private const string DAPR_PUBSUB_NAME = "pubsub";
+
+    private readonly DaprClient _dapr;
+    private readonly ILogger<DaprEventBus> _logger;
+
+    public DaprEventBus(DaprClient dapr, ILogger<DaprEventBus> logger)
+    {
+        _dapr = dapr;
+        _logger = logger;
+    }
+
+    public async Task PublishAsync<TIntegrationEvent>(TIntegrationEvent @event)
+        where TIntegrationEvent : IntegrationEvent
+    {
+        var topicName = @event.GetType().Name;
+
+        _logger.LogInformation("Publishing event {@Event} to {PubsubName}.{TopicName}", @event, DAPR_PUBSUB_NAME, topicName);
+
+        // Make sure to pass the concrete event type to PublishEventAsync,
+        // which can be accomplished by casting the event to dynamic. This ensures
+        // that all event fields are properly serialized.
+        await _dapr.PublishEventAsync(DAPR_PUBSUB_NAME, topicName, (dynamic)@event);
+    }
+}
+```
+
+As you can see in the code snippet, the topic name is derived from event type's name. Because all eShop services use the `IEventBus` abstraction, retrofitting Dapr required *absolutely no change* to the mainline application code.
+
+> Note how we cast the event parameter to a C# `dynamic` type. This workaround is necessary because in .NET 3.1 the `System.Text.Json` serializer does not support polymorphism yet. The Dapr SDK uses `System.Text.Json` to serialize/deserialize messages. In the eShop code, an event is sometimes explicitly declared as an `IntegrationEvent`, the base-class for integration events. This is done because the specific event-type is determined dynamically at runtime based on business-logic. As a result, the event is serialized using the type information of the base-class and not the derived class. The message payload won't contain all the fields of the event, but only those declared on the base-class. Our workaround circumvents this and all the fields are correctly serialized.
```

Comment:
```suggestion
> [!IMPORTANT]
> There is a cast on the event parameter to a C# `dynamic` type. This workaround is necessary because in .NET 3.1 the `System.Text.Json` serializer does not support polymorphism yet. The Dapr SDK uses `System.Text.Json` to serialize/deserialize messages. In the eShop code, an event is sometimes explicitly declared as an `IntegrationEvent`, the base-class for integration events. This is done because the specific event-type is determined dynamically at runtime based on business-logic. As a result, the event is serialized using the type information of the base-class and not the derived class. The message payload won't contain all the fields of the event, but only those declared on the base-class. Our workaround circumvents this and all the fields are correctly serialized.
```